Support http2.

sogaani opened the pull request:
Hi.

This PR enable http/2, if EXPOSE_HTTP2 env is true.

This PR still not support ALPN(as @kornelski mentioned in https://github.com/visionmedia/superagent/issues/980).
I'd prefer to add ALPN functionality after merging this PR and [express http/2 PR](https://github.com/expressjs/express/pull/3390).

kornelski commented:
Thank you! That's a very useful functionality.

kornelski commented:
Does `EXPOSE_HTTP2` affect ability to make HTTP/1 requests? Is Node going to fall back automatically? I wonder if we need to give more runtime control over this.

kornelski commented on lib/node/http2wrapper.js:
```diff
@@ -0,0 +1,186 @@
+  this._headers = {};
+
+  const session = http2.connect(`${protocol}://${host}:${port}`, sessionOptions);
+  this.setHeader('host', `${host}:${port}`)
```
Is it necessary here? In normal implementation we allow `options.host` to be different than host header, because it's useful to do `request.get('http://[ip address]').host('example.com')` to force request to connect to a particular IP address.

sogaani replied:
I think current implementation supports `request.get('http://[ip address]').host('example.com')`. I will check later 👍 .

sogaani replied:
Still not send header here and we can change `Host` header by `set` method. I add test for this.

kornelski commented on lib/node/http2wrapper.js:
```diff
@@ -0,0 +1,186 @@
+    case 'http':
+      return net.connect(options.socketPath);
+    case 'https':
+      options.ALPNProtocols = ['h2'];
```
AFAIK browsers also send http/1 here. Do we support a fallback?

sogaani replied:
Current implement does not support fall back. I can implement fallback, I plan to take fallback after this PR. Or Should I integrate  with this PR?

kornelski replied:
I see. If there's no fallback, then that's OK.

kornelski commented on lib/node/http2wrapper.js:
```diff
@@ -0,0 +1,186 @@
+    switch (key) {
+      case HTTP2_HEADER_HOST:
+        key = HTTP2_HEADER_AUTHORITY;
+        value = value.startsWith('http') ? parse(value).host : value;
```
Why is such conditional parsing needed?

sogaani replied:
This is add for compatibility with Http/1. HTTP2_HEADER_AUTHORITY(alternative Host in http/2) only support domain string, otherwise it raise error. However, I met Host header with schema(like Host: http://example.com) in some express tests.

kornelski replied:
So it would be better to improve the check, otherwise host like `httparchive.org` would be misparsed :)

```js
/^https?:\/\//.test(value)
```

kornelski commented on test/node/pipe.js:
```diff
@@ -12,13 +17,19 @@ app.get("/", (req, res) => {
 app.post("/", (req, res) => {
-  res.send(req.body);
+  if (process.env.EXPOSE_HTTP2){
+    res.set('content-type', 'application/json');
```
Why is this needed?

sogaani replied:
http/2 module have not full compatibility with http/1 module. http/2 imcommingHttpRequest does not have body property.

kornelski commented on test/node/unix-sockets.js:
```diff
@@ -94,6 +105,7 @@ describe("[unix-sockets] https", () => {
         .get(`${base}/request/path`)
         .ca(cert)
         .end((err, res) => {
+          console.log(err);
```
👀

sogaani replied:
My misstake! I will fix 👍

kornelski commented on test/node/user-agent.js:
```diff
@@ -6,7 +6,7 @@ const base = setup.uri;
 describe("req.get()", () => {
   it("should set a default user-agent", () =>
-    request.get(`${base}/ua`).then(res => {
+    request.put(`${base}/ua`).then(res => {
```
Can you explain this change?

sogaani replied:
http/2 client module cannot send body with head or get Method. But, it seems to not need here. I check this again later.

sogaani replied:
I test locally and noticed that this change is not needed. I will revert this change.

kornelski commented on test/support/server.js:
```diff
@@ -17,7 +42,15 @@ app.all('/url', (req, res) => {
+  if(process.env.EXPOSE_HTTP2){
+    Object.keys(headers).forEach(function(name) {
+      if(isPseudoHeader(name)){
+        delete headers[name];
```
Maybe this should translate the header to HTTP/1 header, or allow both in the test using it?

sogaani replied:
HTTP/2 use some headers to control protocol(e.g. :method, :authority, :path ...) and it can not set by user. I just delete those headers here.

kornelski commented:
@focusaurus do you have a preferences how HTTP/1 vs 2 API should look like?

focusaurus commented:
Hmm. Am I interpreting the discussion correctly? So if a user sets `EXPOSE_HTTP2=true` they can only connect to http2 servers from that entire node process and connections to http/1 servers will fail? That sounds untenable. I think I'd prefer leaving HTTP/1 as the default and exposing an API to use http/2 on a per-request basis. For that I think the `.http2()` method name is a good API.

sogaani commented:
@focusaurus Thank you for feedback!.
>  I think I'd prefer leaving HTTP/1 as the default and exposing an API to use http/2 on a per-request basis. For that I think the .http2() method name is a good API.

It is reasonable. I change this soon. 👍

sogaani commented:
I addressed review and change the API to exposing an HTTP/2 API.

Superagent can be enabled http2 by http2 field:

```js
    const request = require('superagent');
    request.http2 =true;
    request
      .get('http://example.com/search')
      .then(res => {

      });
```

A request can be enabled http2 by http2 method:

```js
    const request = require('superagent');
    request
      .get('http://example.com/search')
      .http2()
      .then(res => {

      });
```

sogaani commented:
ping @kornelski  @focusaurus

kornelski commented:
Thanks

focusaurus commented:
Awesome contribution @sogaani !
